### PR TITLE
[4.0] RFC Quickicons Joomla and Extensions updates

### PR DESCRIPTION
**Pull request by infograf768**
Replaces and complete https://github.com/joomla/joomla-cms/pull/23061 (for what concerns Joomla Update).

Implement NOTE in https://github.com/joomla/joomla-cms/pull/23061#issuecomment-438581619

### Summary of Changes
1. Adding Joomla Version
2. Change icon color to `warning` when an update is available as `danger` is used in case of error.
3. Place the `Update Now` and `View Updates` buttons to the bottom right of the Message, centering the rest of the message.
4. Adapt Atum CSS concerned to RTL, including the `joomla-alert`
5. Correct unnecessary `,` and `:` in the language strings
5. Correct `Text::script` to `false` by default to let single quotes in language strings to display unescaped instead of `\'`

### Testing Instructions
_To test on 4.0-dev branch_

**Concerning joomlaupdate:**
Modify version.php line 63 to const EXTRA_VERSION = 'alpha4-dev';
Go to database and Fix database.
Then in JoomlaUpdate Options choose Custom URL and add this link (Thanks Tobias) :
https://update.joomla.org/core/nightlies/next_major_list.xml
Save Options.
Display Control Panel.

**Concerning extensions update:**
Install an older language pack in order to get a warning for an update.
[fr-FR_joomla_lang_full_3.9.0v1.zip](https://github.com/joomla/joomla-cms/files/2570924/fr-FR_joomla_lang_full_3.9.0v1.zip)

in both cases, test also RTL by installing and switching to Persian language.

Test before patch, then run npm ci

### Expected result
screen shot 2018-11-16 at 11 22 24
screen shot 2018-11-16 at 09 31 19
screen shot 2018-11-16 at 09 30 00
screen shot 2018-11-16 at 11 45 28

@tuum @wilsonge

### CODE REVIEW

File: administrator/templates/atum/scss/template-rtl.scss [hunk 1]
```diff
@@ -249,7 +249,20 @@ body {
...
 }
 
-
 .field-media-input {
   margin-right: 220px;
 }
+
+// Quickicon specific
+.message-alert {
+  text-align: left !important;
```
Comment (brianteeman) [inline administrator/templates/atum/scss/template-rtl.scss:258]: surely there is a way to do this without using !important

File: administrator/language/en-GB/en-GB.plg_quickicon_joomlaupdate.ini [hunk 2]
```diff
@@ -8,8 +8,8 @@ PLG_QUICKICON_JOOMLAUPDATE_CHECKING="Checking Joomla ..."
 PLG_QUICKICON_JOOMLAUPDATE_ERROR="Unknown Joomla ..."
 PLG_QUICKICON_JOOMLAUPDATE_GROUP_DESC="The group of this plugin (this value is compared with the group value used in <strong>Quick Icons</strong> modules to inject icons)."
 PLG_QUICKICON_JOOMLAUPDATE_GROUP_LABEL="Group"
-PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND="Joomla %s, Update now!"
+PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND="Joomla %s Update now!"
 PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND_BUTTON="Update Now"
-PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND_MESSAGE="Joomla %s is available:"
+PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND_MESSAGE="Joomla %s is available"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_quickicon_joomlaupdate.ini]: This is now a sentence so should end with a full stop.

```suggestion
PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND_MESSAGE="Joomla %s is available."
```

File: administrator/language/en-GB/en-GB.plg_quickicon_extensionupdate.ini [hunk 3]
```diff
@@ -10,6 +10,6 @@ PLG_QUICKICON_EXTENSIONUPDATE_GROUP_DESC="The group of this plugin (this value i
 PLG_QUICKICON_EXTENSIONUPDATE_GROUP_LABEL="Group"
 PLG_QUICKICON_EXTENSIONUPDATE_UPDATEFOUND="Updates are available! %s"
 PLG_QUICKICON_EXTENSIONUPDATE_UPDATEFOUND_BUTTON="View Updates"
-PLG_QUICKICON_EXTENSIONUPDATE_UPDATEFOUND_MESSAGE="%s Extension Update(s) are available:"
+PLG_QUICKICON_EXTENSIONUPDATE_UPDATEFOUND_MESSAGE="%s Extension Update(s) are available"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_quickicon_extensionupdate.ini]: This is now a sentence so should end with a full stop.

```suggestion
PLG_QUICKICON_JOOMLAUPDATE_UPDATEFOUND_MESSAGE="Joomla %s is available."
```

**Review by brianteeman — CHANGES_REQUESTED** [review]
punctuation as commented inline

File: administrator/templates/atum/scss/template-rtl.scss (hunk 1 above)
Comment (infograf768) [reply administrator/templates/atum/scss/template-rtl.scss:258]: This is what is used in node_modules and vendor:
```
@each $breakpoint in map-keys($grid-breakpoints) {
  @include media-breakpoint-up($breakpoint) {
    $infix: breakpoint-infix($breakpoint, $grid-breakpoints);

    .text#{$infix}-left   { text-align: left !important; }
    .text#{$infix}-right  { text-align: right !important; }
    .text#{$infix}-center { text-align: center !important; }
  }
}
```
and ported to `/administrator/templates/atum/css/bootstrap.css` once npm ci has run:
```
.text-left {
  text-align: left !important; }

.text-right {
  text-align: right !important; }

.text-center {
  text-align: center !important; }
```

**Review by brianteeman — APPROVED** [review]
approve language strings

File: administrator/templates/atum/scss/template-rtl.scss [hunk 4]
```diff
@@ -249,7 +249,20 @@ body {
...
+}
+
+// Message specific
+joomla-alert .joomla-alert--close, joomla-alert .joomla-alert-button--close {
+    right: 0;
```
Comment (Quy) [inline administrator/templates/atum/scss/template-rtl.scss]: Use 2 spaces to indent.

Comment (infograf768) [reply administrator/templates/atum/scss/template-rtl.scss]: done.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on a41940342df38346ee358f349e4c513e8af1e995<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

**Comment by infograf768** [comment]
Note: furher PR will correct RTL for the J version

**Comment by infograf768** [comment]
@richard67 
I need to update this now for RTL version but also for another js issue.
Please test again after I updated.

**Comment by infograf768** [comment]
@richard67 
OK, you can test again now.
The modifications concern the version in RTL (Test by using Persian), as well as showing the Joomlaupdate Quickicon in Green when there is no Update available.

screen shot 2018-11-18 at 11 40 31
screen shot 2018-11-18 at 11 48 41

**Comment by richard67** [comment]
@infograf768 I don't see any difference in the Persian screenshots. What has changed?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

**Comment by infograf768** [comment]
formerly (see on top), the version read alpha6-dev etc.
it now displays correctly 4.0.0-alpha etc, i.e. the version is ltr in a rtl lang string

**Comment by richard67** [comment]
I see now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

**Comment by richard67** [comment]
@infograf768 Now I tested and it seemed ok. But after switching back the Joomla update channel to "Default" and changing back Version.php and fixing the database, I get a javascript error from the Joomla update check, and the update check does not finish:
screen shot 2018-11-18 at 11 55 18
Reverting the PR and running "npm ci" does not fix that, so I am not sure if related to this PR or not.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

**Comment by richard67** [comment]
@infograf768 I've started again with clean 4.0-dev, installed patch tester, applied your patch, run "npm ci", switched on system debugging in global settings so uncompressed js is used.
The js error is in jupdatecheck.js, row 32, column 11, which is statement "if (updateInfo.version === options.version) {", and sais "TypeError: updateInfo is undefined".
This seems to come from your last change.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

**Comment by richard67** [comment]
@infograf768 It seems that the "if (updateInfoList.length === 0) {" you've removed with your last change was not there for no reason.

**Comment by infograf768** [comment]
on ipad now. will look later. the issue was to get the ‘success’ class instead of only ‘pulse’. we may have to define as var and empty the ‘updateinfo.version’ first.

**Comment by richard67** [comment]
@infograf768 Can you check if my PR [https://github.com/infograf768/joomla-cms/pull/50](https://github.com/infograf768/joomla-cms/pull/50) against your branch is the right thing to correct it and get what you want with the success class? Unfortunately I made a mistake in title of the commit so it contains "warning" instead of "success", which is misleading. So you better just check it and if ok do the same change yourself.

**Comment by richard67** [comment]
@infograf768 I've just tested my changes from [https://github.com/infograf768/joomla-cms/pull/50](https://github.com/infograf768/joomla-cms/pull/50), applied to this PR here, and it works. Javascript error is gone, and Joomla Update is green.

**Comment by infograf768** [comment]
Thanks @richard67 
I mark your test Ok again.

**Comment by richard67** [comment]
Yes, I've just tested again, works all well.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

File: administrator/templates/atum/scss/template-rtl.scss [hunk 5]
```diff
@@ -249,7 +249,20 @@ body {
...
   margin-right: 220px;
 }
+
+// Quickicon specific
+.message-alert {
+  text-align: left !important;
+}
+
+// Message specific
+joomla-alert .joomla-alert--close, joomla-alert .joomla-alert-button--close {
```
Comment (C-Lodder) [inline administrator/templates/atum/scss/template-rtl.scss]: Alerts are not part of the template. please put RTL CSS in the custom element file

Comment (infograf768) [reply administrator/templates/atum/scss/template-rtl.scss]: @C-Lodder
This is an override for RTL.
The scss for alert is in `/node_modules/joomla-ui-custom-elements/src/scss/alert/alert.scss`

There is no provisions done for RTL in custom-elements which means we would have to use `html[dir=rtl] `(which is not a problem I guess).

Also it means that until someone merges a possible pull request in https://github.com/joomla-projects/custom-elements/blob/master/dist/css/joomla-alert.css and possibly https://github.com/joomla-projects/custom-elements/blob/master/dist/css/joomla-alert.min.css, this PR would not correct by itself the RTL display issue with alerts.

EDIT: Maybe it is only in https://github.com/joomla-projects/custom-elements/blob/master/src/scss/alert/alert.scss

The other solution I see is to add these (with `html[dir=rtl] `), in the `atum/scss/blocks/_alert.scss` and also in Cassiopea equivalent file —which means that every 3rd party template will necessarily have to do that to be rtl compatible.

Just let me know what is the way to go.
calling @wilsonge as he has the hand on merging stuff in https://github.com/joomla-projects/custom-elements

Comment (infograf768) [reply administrator/templates/atum/scss/template-rtl.scss]: @C-Lodder 
ping 😄

Comment (infograf768) [reply administrator/templates/atum/scss/template-rtl.scss]: After discussion with @wilsonge will make PR towards https://github.com/joomla-projects/custom-elements/blob/master/src/scss/alert/alert.scss
and delete the same in template-rtl.scss>

We will have to wait the PR is merged and compile is done to see the rtl results here.

**Comment by infograf768** [comment]
RTL can now be tested fully as alert has been updated.
One has to use present 4.0.0-dev branch, patch and run npm ci.

Result:
screen shot 2018-11-23 at 10 12 00

@richard67

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 4ed7a1bc0d797c4491e1c95d6a680770bb4b2f01<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23095">issues.joomla.org/tracker/joomla-cms/23095</a>.</sub>

**Comment by infograf768** [comment]
@wilsonge
same comment as in the other PR.
RTL testers are a rare category... ;)

**Comment by wilsonge** [comment]
Thanks JM!

**Comment by infograf768** [comment]
and @richard67  ;)

**Pull request merged**